### PR TITLE
feat: support athenz authentication for pulsar

### LINKED ISSUE #2558: support athenz authentication for pulsar
**Is your feature request related to a problem? Please describe.**
I want to use a pulsar eventsource/sensor with athenz authentication which is described in [docs](https://pulsar.apache.org/docs/2.11.x/security-athenz/).

**Describe the solution you'd like**
Support athenz authentication for pulsar eventsource/sensor.
Client library pulsar-client-go using in argo events have already supported it.
https://github.com/apache/pulsar-client-go/blob/75d2df3b7d1d1d04fb660a1b6c11ede1d2f161bf/pulsar/client.go#L70-L74

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by khmjp**
fix #2558
In eventsource and trigger, added athenz objects to apis and process.

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-events/blob/master/USERS.md).

### CODE REVIEW

File: pkg/apis/eventsource/v1alpha1/types.go [hunk 1]
```diff
@@ -1374,11 +1374,21 @@ type PulsarEventSource struct {
...
 	Metadata map[string]string `json:"metadata,omitempty" protobuf:"bytes,10,rep,name=metadata"`
 	// Authentication token for the pulsar client.
+	// Either token or athenz can be set to use auth.
 	// +optional
 	AuthTokenSecret *corev1.SecretKeySelector `json:"authTokenSecret,omitempty" protobuf:"bytes,11,opt,name=authTokenSecret"`
+	// Authentication athenz parameters for the pulsar client.
+	// Refer https://github.com/apache/pulsar-client-go/blob/master/pulsar/auth/athenz.go
+	// Either token or athenz can be set to use auth.
+	// +optional
+	AuthAthenzParams map[string]string `json:"authAthenzParams,omitempty" protobuf:"bytes,12,opt,name=authAthenzParams"`
```
Comment (whynowy) [inline pkg/apis/eventsource/v1alpha1/types.go:1384]: `bytes,12,rep,name=xxx`

File: pkg/apis/eventsource/v1alpha1/types.go [hunk 2]
```diff
@@ -1374,11 +1374,21 @@ type PulsarEventSource struct {
...
+	// Authentication athenz privateKey secret for the pulsar client.
+	// AuthAthenzSecret must be set if AuthAthenzParams is used.
+	// +optional
+	AuthAthenzSecret *corev1.SecretKeySelector `json:"authAthenzSecret,omitempty" protobuf:"bytes,13,opt,name=authAthenzSecret"`
 	// Filter
 	// +optional
-	Filter *EventSourceFilter `json:"filter,omitempty" protobuf:"bytes,12,opt,name=filter"`
+	Filter *EventSourceFilter `json:"filter,omitempty" protobuf:"bytes,14,opt,name=filter"`
```
Comment (whynowy) [inline pkg/apis/eventsource/v1alpha1/types.go:1391]: Can you keep the previous index, and add the 2 new in the end?

File: pkg/apis/sensor/v1alpha1/types.go [hunk 3]
```diff
@@ -636,11 +636,21 @@ type PulsarTrigger struct {
...
 	TLS *apicommon.TLSConfig `json:"tls,omitempty" protobuf:"bytes,8,opt,name=tls"`
 	// Authentication token for the pulsar client.
+	// Either token or athenz can be set to use auth.
 	// +optional
 	AuthTokenSecret *corev1.SecretKeySelector `json:"authTokenSecret,omitempty" protobuf:"bytes,9,opt,name=authTokenSecret"`
+	// Authentication athenz parameters for the pulsar client.
+	// Refer https://github.com/apache/pulsar-client-go/blob/master/pulsar/auth/athenz.go
+	// Either token or athenz can be set to use auth.
+	// +optional
+	AuthAthenzParams map[string]string `json:"authAthenzParams,omitempty" protobuf:"bytes,10,opt,name=authAthenzParams"`
```
Comment (whynowy) [inline pkg/apis/sensor/v1alpha1/types.go:646]: `rep`

File: pkg/apis/sensor/v1alpha1/types.go [hunk 4]
```diff
@@ -636,11 +636,21 @@ type PulsarTrigger struct {
...
+	// Authentication athenz privateKey secret for the pulsar client.
+	// AuthAthenzSecret must be set if AuthAthenzParams is used.
+	// +optional
+	AuthAthenzSecret *corev1.SecretKeySelector `json:"authAthenzSecret,omitempty" protobuf:"bytes,11,opt,name=authAthenzSecret"`
 	// Backoff holds parameters applied to connection.
 	// +optional
-	ConnectionBackoff *apicommon.Backoff `json:"connectionBackoff,omitempty" protobuf:"bytes,10,opt,name=connectionBackoff"`
+	ConnectionBackoff *apicommon.Backoff `json:"connectionBackoff,omitempty" protobuf:"bytes,12,opt,name=connectionBackoff"`
```
Comment (whynowy) [inline pkg/apis/sensor/v1alpha1/types.go:653]: Same here for the index.

File: sensors/triggers/pulsar/pulsar.go [hunk 5]
```diff
@@ -73,6 +73,17 @@ func NewPulsarTrigger(sensor *v1alpha1.Sensor, trigger *v1alpha1.Trigger, pulsar
 			clientOpt.Authentication = pulsar.NewAuthenticationToken(token)
 		}
 
+		if pulsarTrigger.AuthAthenzParams != nil && pulsarTrigger.AuthAthenzSecret != nil {
+			logger.Info("setting athenz auth option...")
+			authAthenzFilePath, err := common.GetSecretVolumePath(pulsarTrigger.AuthAthenzSecret)
+			if err != nil {
+				logger.Errorw("failed to get authAthenzSecret from the volume", "error", err)
```
Comment (whynowy) [inline sensors/triggers/pulsar/pulsar.go:80]: `zap.Error(err)`

File: eventsources/sources/pulsar/start.go [hunk 6]
```diff
@@ -107,6 +107,17 @@ func (el *EventListener) StartListening(ctx context.Context, dispatch func([]byt
 		clientOpt.Authentication = pulsar.NewAuthenticationToken(token)
 	}
 
+	if pulsarEventSource.AuthAthenzParams != nil && pulsarEventSource.AuthAthenzSecret != nil {
```
Comment (whynowy) [inline eventsources/sources/pulsar/start.go:110]: Is there a better way to tell if the `AuthAthenzParams` is correctly configured? e.g. if it contains a required field?

If there's not a better way, I suggest to check it with `len(pulsarEventSource.AuthAthenzParams) > 0`.

Comment (khmjp) [reply eventsources/sources/pulsar/start.go:110]: Thanks for the feedback. I will gratefully accept your proposal.

**Review by whynowy — APPROVED** [review]
LGTM, thanks!

**Pull request merged**